### PR TITLE
James-3317  Configuration option: choosing between PassThroughBlobStore & DeDuplicatingBlobStore

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/blob.properties [hunk 1]
```diff
@@ -6,6 +6,13 @@
 # Optional, default is localFile
 blob.export.implementation=localFile
 
+# ========================================= ObjectStorage deduplication ========================================
+# If you choose to enable deduplication, the mails with the same content will be stored only once.
+# Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-ldap/destination/conf/blob.properties:11]:
```suggestion
# Warning: Once this feature is enabled, there is no turning back as turning it off will lead to the deletion of all
```

File: dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/blob.properties [hunk 2]
```diff
@@ -5,6 +5,13 @@
 # Mandatory, allowed values are: cassandra, objectstorage
 implementation=objectstorage
 
+# ========================================= ObjectStorage deduplication ========================================
+# If you choose to enable deduplication, the mails with the same content will be stored only once.
+# Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/blob.properties:10]: idem

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/blob.properties [hunk 3]
```diff
@@ -5,6 +5,13 @@
 # Mandatory, allowed values are: cassandra, objectstorage
 implementation=objectstorage
 
+# ========================================= ObjectStorage deduplication ========================================
+# If you choose to enable deduplication, the mails with the same content will be stored only once.
+# Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/blob.properties:10]: idem

File: dockerfiles/run/guice/cassandra/destination/conf/blob.properties [hunk 4]
```diff
@@ -6,6 +6,13 @@
 # Optional, default is localFile
 blob.export.implementation=localFile
 
+# ========================================= ObjectStorage deduplication ========================================
+# If you choose to enable deduplication, the mails with the same content will be stored only once.
+# Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra/destination/conf/blob.properties:11]: idem

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreConfiguration.java [hunk 5]
```diff
@@ -114,30 +149,44 @@ static BlobStoreConfiguration from(Configuration configuration) {
                 "supported values in: %s", BLOBSTORE_IMPLEMENTATION_PROPERTY, BlobStoreImplName.supportedImplNames())));
 
         boolean cacheEnabled = configuration.getBoolean(CACHE_ENABLE_PROPERTY, false);
-
-        return new BlobStoreConfiguration(blobStoreImplName, cacheEnabled);
+        boolean deduplicationEnabled = Try.ofCallable(() -> configuration.getBoolean(DEDUPLICATION_ENABLE_PROPERTY))
+                .getOrElseThrow(() -> new IllegalStateException("deduplication.enable property is missing please use one of the supported values in: true, false\n" +
+                        "If you choose to enable deduplication, the mails with the same content will be stored only once.\n" +
+                        "Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all\n" +
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreConfiguration.java:155]: idem

File: server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/modules/blobstore/BlobStoreConfigurationTest.java [hunk 6]
```diff
@@ -223,8 +239,43 @@ void cacheEnabledShouldBeFalseWhenSpecified() {
...
+
+    @Test
+    void buildingConfigurationShouldThrowWhenDeduplicationPropertieIsOmitted() {
+        PropertiesConfiguration configuration = new PropertiesConfiguration();
+        configuration.addProperty("implementation", BlobStoreConfiguration.BlobStoreImplName.OBJECTSTORAGE.getName());
+
+        assertThatThrownBy(() -> BlobStoreConfiguration.from(configuration)).isInstanceOf(IllegalStateException.class)
+                .hasMessage("deduplication.enable property is missing please use one of the supported values in: true, false\n" +
+         "If you choose to enable deduplication, the mails with the same content will be stored only once.\n" +
+         "Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all\n" +
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/modules/blobstore/BlobStoreConfigurationTest.java:276]: idem

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 7]
```diff
@@ -38,22 +40,43 @@
...
 
             bind(DumbBlobStore.class).to(CassandraDumbBlobStore.class);
 
-            bind(BlobStore.class)
-                .annotatedWith(Names.named(CachedBlobStore.BACKEND))
-                .to(DeDuplicationBlobStore.class);
+            if (StorageStrategy.DEDUPLICATION == storageStrategy) {
+                bind(BlobStore.class)
+                        .annotatedWith(Names.named(CachedBlobStore.BACKEND))
+                        .to(DeDuplicationBlobStore.class);
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:58]: indent

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 8]
```diff
@@ -38,22 +40,43 @@
...
+            } else if (StorageStrategy.PASSTHROUGH == storageStrategy) {
+                bind(BlobStore.class)
+                        .annotatedWith(Names.named(CachedBlobStore.BACKEND))
+                        .to(PassThroughBlobStore.class);
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:62]: indent

File: src/site/xdoc/server/config-blobstore.xml [hunk 9]
```diff
@@ -46,6 +46,13 @@
                 <dt><strong>implementation</strong></dt>
                 <dd>cassandra: use cassandra based BlobStore</dd>
                 <dd>objectstorage: use Swift/AWS S3 based BlobStore</dd>
+
+                <dt><strong>deduplication/enable</strong></dt>
+                <dd>Supported value: true and false. (default: false)</dd>
+                <dd>If you choose to enable deduplication, the mails with the same content will be stored only once.</dd>
+                <dd>Warning: Once this feature is enabled, there is no turning back has turning it off will lead to deletion of all</dd>
```
Comment (Arsnael) [inline src/site/xdoc/server/config-blobstore.xml:53]: This too

**Comment by remk** [comment]
```
[862b4f6cd6a3c458838a95ef082f162883b69f29] [ERROR] Tests run: 85, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 13.408 s <<< FAILURE! - in org.apache.james.webadmin.vault.routes.DeletedMessagesVaultRoutesTest
[862b4f6cd6a3c458838a95ef082f162883b69f29] [ERROR] exportShouldSaveDeletedMessagesDataToBlobStore  Time elapsed: 0.108 s  <<< ERROR!
[862b4f6cd6a3c458838a95ef082f162883b69f29] org.apache.james.blob.api.ObjectNotFoundException: blob '5f0dc49dd59961732bb0a16c2d8e1042f1c2debd0a01a773a5802cfca81cb99d' not found in bucket 'default'
[862b4f6cd6a3c458838a95ef082f162883b69f29] 	at org.apache.james.webadmin.vault.routes.DeletedMessagesVaultRoutesTest$ExportTest.exportShouldSaveDeletedMessagesDataToBlobStore(DeletedMessagesVaultRoutesTest.java:1938)
[862b4f6cd6a3c458838a95ef082f162883b69f29] 
[862b4f6cd6a3c458838a95ef082f162883b69f29] [INFO] Running org.apache.james.webadmin.vault.routes.WebadminApiQuerySerializationContractTest
[862b4f6cd6a3c458838a95ef082f162883b69f29] [INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.025 s - in org.apache.james.webadmin.vault.routes.WebadminApiQuerySerializationContractTest
[862b4f6cd6a3c458838a95ef082f162883b69f29] [INFO] 
[862b4f6cd6a3c458838a95ef082f162883b69f29] [INFO] Results:
[862b4f6cd6a3c458838a95ef082f162883b69f29] [INFO] 
[862b4f6cd6a3c458838a95ef082f162883b69f29] [ERROR] Errors: 
[862b4f6cd6a3c458838a95ef082f162883b69f29] [ERROR]   DeletedMessagesVaultRoutesTest$ExportTest.exportShouldSaveDeletedMessagesDataToBlobStore:1938 » ObjectNotFound
```

**Comment by chibenwa** [comment]
I fixed it in the base PR ^^

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 10]
```diff
@@ -38,22 +40,43 @@
...
         @Override
         protected void configure() {
             install(new CassandraBlobStoreDependenciesModule());
 
             bind(DumbBlobStore.class).to(CassandraDumbBlobStore.class);
 
-            bind(BlobStore.class)
-                .annotatedWith(Names.named(CachedBlobStore.BACKEND))
-                .to(DeDuplicationBlobStore.class);
+            if (StorageStrategy.DEDUPLICATION == storageStrategy) {
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:55]: DOn't do IF in guice module, do if in the coose method!

**Comment by chibenwa** [comment]
Forced pushed to remove the no review commit of the passthrough blobStore

**Comment by chibenwa** [comment]
```
[88d6c2845fdb354a811ff0773ac0c01401c0c861] 20:40:44.083 [ERROR] o.a.j.GuiceJamesServer - Fatal error while starting James
[88d6c2845fdb354a811ff0773ac0c01401c0c861] com.google.inject.CreationException: Unable to create injector, see the following errors:
[88d6c2845fdb354a811ff0773ac0c01401c0c861] 
[88d6c2845fdb354a811ff0773ac0c01401c0c861] 1) No implementation for org.apache.james.blob.api.DumbBlobStore was bound.
[88d6c2845fdb354a811ff0773ac0c01401c0c861]   while locating org.apache.james.blob.api.DumbBlobStore
[88d6c2845fdb354a811ff0773ac0c01401c0c861]     for the 1st parameter of org.apache.james.server.blob.deduplication.DeDuplicationBlobStore.<init>(DeDuplicationBlobStore.scala:41)
[88d6c2845fdb354a811ff0773ac0c01401c0c861]   at org.apache.james.modules.blobstore.BlobStoreModulesChooser.lambda$chooseStoragePolicyModule$0(BlobStoreModulesChooser.java:82) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.blobstore.BlobStoreModulesChooser$$Lambda$534/0x000000010057c840)
[88d6c2845fdb354a811ff0773ac0c01401c0c861] 
[88d6c2845fdb354a811ff0773ac0c01401c0c861] 2) A binding to org.apache.james.blob.api.BlobStore annotated with @com.google.inject.name.Named(value="blobStoreBackend") was already configured at org.apache.james.modules.blobstore.BlobStoreModulesChooser$ObjectStorageDumdBlobStoreDeclarationModule.configure(BlobStoreModulesChooser.java:55) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.blobstore.BlobStoreModulesChooser$ObjectStorageDumdBlobStoreDeclarationModule).
[88d6c2845fdb354a811ff0773ac0c01401c0c861]   at org.apache.james.modules.blobstore.BlobStoreModulesChooser.lambda$chooseStoragePolicyModule$0(BlobStoreModulesChooser.java:82) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.blobstore.BlobStoreModulesChooser$$Lambda$534/0x000000010057c840)
```

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java [hunk 11]
```diff
@@ -37,20 +39,16 @@
...
-                .annotatedWith(Names.named(CachedBlobStore.BACKEND))
-                .to(DeDuplicationBlobStore.class);
         }
     }
 
-    static class ObjectStorageDeclarationModule extends AbstractModule {
+    static class ObjectStorageDumdBlobStoreDeclarationModule extends AbstractModule {
         @Override
         protected void configure() {
             install(new ObjectStorageDependenciesModule());
```
Comment (Arsnael) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:54]: I tried to look a bit into the Guice errors... The problem might possibly coming from here I think?

You need to drop your bind below as it conflicts with what you declared in `chooseStoragePolicyModule` and here I would say you need to bind the  `DumbBlobStore` to a... `ObjectStorageDumdBlobStore` that still not exist it seems (in an other PR? I forgot)

Comment (chibenwa) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:54]: Yes, that class is introduced by the S3 PR, and this work is conditioned to it.

Comment (remk) [reply server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:54]: rebased on #3614 the NoopDumbBlobStore should fix the guice issue for the moment

**Comment by chibenwa** [comment]
`CassandraAccessTokenRepositoryTest>AccessTokenRepositoryTest.validTokenMustBeRetrieved:45 » InvalidAccessToken`

test this please

**Comment by chibenwa** [comment]
```
[24a924a0538b566263ca5bba6fb8b67b8f31c106] 06:50:39.283 [ERROR] o.a.j.GuiceJamesServer - Fatal error while starting James
[24a924a0538b566263ca5bba6fb8b67b8f31c106] com.google.inject.CreationException: Unable to create injector, see the following errors:
[24a924a0538b566263ca5bba6fb8b67b8f31c106] 
[24a924a0538b566263ca5bba6fb8b67b8f31c106] 1) A binding to org.apache.james.blob.api.BlobStore annotated with @com.google.inject.name.Named(value="blobStoreBackend") was already configured at org.apache.james.modules.blobstore.BlobStoreModulesChooser$ObjectStorageDumdBlobStoreDeclarationModule.configure(BlobStoreModulesChooser.java:56) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.blobstore.BlobStoreModulesChooser$ObjectStorageDumdBlobStoreDeclarationModule).
[24a924a0538b566263ca5bba6fb8b67b8f31c106]   at org.apache.james.modules.blobstore.BlobStoreModulesChooser.lambda$chooseStoragePolicyModule$0(BlobStoreModulesChooser.java:83) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.blobstore.BlobStoreModulesChooser$$Lambda$534/0x000000010057c840)
```

**Comment by chibenwa** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (default) on project james-server-cassandra-rabbitmq-guice: You have 1 Checkstyle violation. -> [Help 1]

[ERROR] src/main/java/org/apache/james/modules/blobstore/BlobStoreModulesChooser.java:[68,9] (whitespace) WhitespaceAround: WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)
```

**Comment by chibenwa** [comment]
Merged

**Review by mbaechler — COMMENTED** [review]
If I understand correctly, we don't prevent the admin to loose data by switching the deduplication property after server initialization, that's it?

**Comment by rouazana** [comment]
> If I understand correctly, we don't prevent the admin to loose data by switching the deduplication property after server initialization, that's it?

wrong, it has been handled here https://github.com/linagora/james-project/issues/3596 via this PR: https://github.com/linagora/james-project/pull/3611

**Comment by mbaechler** [comment]
> > If I understand correctly, we don't prevent the admin to loose data by switching the deduplication property after server initialization, that's it?
> 
> wrong, it has been handled here #3596 via this PR: #3611

cool

**Pull request closed without merging**